### PR TITLE
ndimage: complex dtype support in filters and interpolation

### PR DESCRIPTION
**Pull request by grlee77**
~blocked by #4402~

This PR enables interpolation (scipy/scipy#12812) and convolution/correlation of complex-valued data (scipy/scipy#12725). It is created starting from #4401, so most commits here are duplicate with that PR. The new ones start from 87872b7.

Work on this PR resulted in a bug fix PR to SciPy 1.6.0rc1 (scipy/scipy#13249) for constant boundary modes with `cval` != 0.

### CODE REVIEW

**Comment by coderforlife** [comment]
A few questions:

- Does this add `"grid-*"` mode to all `ndimage` functions (although they will be synonymous with other modes)?
- Scipy's version of `ndimage.filter` functions originally required all temporaries to be `float64`, is this still true with the additional requirement that `complex` inputs end up using `complex128` temporaries? (see #3705)
- (maybe not answerable yet) Do the changes here mean that all of the uses of the internal `filter` functions in the `signal` library can start supporting complex numbers? That would be great! (current functions which have intentionally blocked complex numbers are `signal.convolve`, `signal.correlate`, `signal.convolve2d`, `signal.correlate2d`, `signal.wiener`, and `signal.sepfir2d` - first four go through `_get_correlate_kernel`, last two through `uniform_filter` and `_run_1d_correlates`).

Otherwise really looking forward to this!

**Comment by grlee77** [comment]
> Does this add "grid-*" mode to all ndimage functions (although they will be synonymous with other modes)?

That is the intention, but unfortunately SciPy 1.6.0rc1 will segfault if  'grid-constant' is used with `ndimage.filters` functions. It should be resolved by scipy/scipy#13249

> Scipy's version of ndimage.filter functions originally required all temporaries to be float64, is this still true with the additional requirement that complex inputs end up using complex128 temporaries? (see #3705)

Everything is still float64 temporaries in SciPy. Unlike this PR where complex dtypes are used in the kernels, in SciPy the C code is just [run independently on complex and real components](https://github.com/scipy/scipy/blob/3cc683418c30b00f2d4af242d010d969b3527ba1/scipy/ndimage/filters.py#L54-L71). Otherwise adding this to SciPy would have been a much bigger effort to overhaul the full ndimage C code.

> Do the changes here mean that all of the uses of the internal filter functions in the signal library can start supporting complex numbers?

In CuPy, I think so. All ndimage convolution/correlation based filters should work for with complex-valued images and/or kernel weights.

In SciPy the convolutions under `scipy.signal` unfortunately use an entirely independent C code base from ndimage, so I think an upstream PR for this would still be needed.

**Comment by coderforlife** [comment]
Interesting that it just does real and imaginary parts separately. When I was referring to the `signal` functions that need complex support, they already support it in scipy, but not in cupy. Hopefully once this PR is committed then I can make all of those functions support complex as well by just removing some if statements.

**Comment by takagi** [comment]
Now all the proceeding PRs are merged to the master! Would you resolve the conflicts?

**Comment by grlee77** [comment]
This has now been rebased. Thanks for getting all of these recent ndimage PRs in!

File: cupyx/scipy/ndimage/_filters_core.py [hunk 1]
```diff
@@ -174,6 +174,10 @@ def _call_kernel(kernel, input, weights, output, structure=None,
 template <class T>
 __device__ __forceinline__ bool nonzero(T x) { return x!=0; }
 
+template <typename T>
+__device__ __forceinline__ bool nonzero(complex<T> x) {
+    return x.real() || x.imag();
+}
```
Comment (leofang) [inline cupyx/scipy/ndimage/_filters_core.py:180]: I thought this is not necessary? `x!=0` is a valid comparison even for `x` being complex.

Comment (grlee77) [reply cupyx/scipy/ndimage/_filters_core.py:180]: I think I was getting compilation failures until I added this back in, but let me try again now without it and see.

Comment (grlee77) [reply cupyx/scipy/ndimage/_filters_core.py:180]: If I remove this nonzero case, and then run
`pytest tests/cupyx_tests/scipy_tests/ndimage_tests/test_filters.py -k TestFilterComplexFast -x`

I get the following error:
```
tests/cupyx_tests/scipy_tests/ndimage_tests/test_filters.py F

============================================================= FAILURES ==============================================================
_ TestFilterComplexFast.test_filter[_param_0_{axis=0, dtype=complex64, filter='gaussian_filter1d', output=None, shape=(4, 5), sigma=1.5, truncate=2.75}] _

self = <<cupyx_tests.scipy_tests.ndimage_tests.test_filters.TestFilterComplexFast object at 0x7f386981d358>  parameter: {'axi...ussian_filter1d', 'sigma': 1.5, 'truncate': 2.75, 'dtype': <class 'numpy.complex64'>, 'output': None, 'shape': (4, 5)}>
xp = <module 'cupy' from '/home/lee8rx/my_git/pyir/cupy/cupy/__init__.py'>
scp = <module 'cupyx.scipy' from '/home/lee8rx/my_git/pyir/cupy/cupyx/scipy/__init__.py'>

    @testing.numpy_cupy_allclose(atol=1e-5, rtol=1e-5, scipy_name='scp')
    @testing.with_requires('scipy>=1.6.0')
    def test_filter(self, xp, scp):
>       return self._filter(xp, scp)

tests/cupyx_tests/scipy_tests/ndimage_tests/test_filters.py:323: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/cupyx_tests/scipy_tests/ndimage_tests/test_filters.py:89: in _filter
    return filter(arr, *args, **kwargs)
cupyx/scipy/ndimage/filters.py:320: in gaussian_filter1d
    return correlate1d(input, weights, axis, output, mode, cval)
cupyx/scipy/ndimage/filters.py:113: in correlate1d
    return _correlate_or_convolve(input, weights, output, mode, cval, origins)
cupyx/scipy/ndimage/filters.py:177: in _correlate_or_convolve
    weights_dtype=weights_dtype)
cupyx/scipy/ndimage/_filters_core.py:141: in _call_kernel
    kernel(*args)
cupy/core/_kernel.pyx:821: in cupy.core._kernel.ElementwiseKernel.__call__
    kern = self._get_elementwise_kernel(dev_id, arginfos, type_map)
cupy/core/_kernel.pyx:846: in cupy.core._kernel.ElementwiseKernel._get_elementwise_kernel
    kern = _get_elementwise_kernel(
cupy/_util.pyx:59: in cupy._util.memoize.decorator.ret
    result = f(*args, **kwargs)
cupy/core/_kernel.pyx:639: in cupy.core._kernel._get_elementwise_kernel
    return _get_simple_elementwise_kernel(
cupy/core/_kernel.pyx:37: in cupy.core._kernel._get_simple_elementwise_kernel
    cpdef function.Function _get_simple_elementwise_kernel(
cupy/core/_kernel.pyx:60: in cupy.core._kernel._get_simple_elementwise_kernel
    module = compile_with_cache(module_code, options)
cupy/core/core.pyx:1986: in cupy.core.core.compile_with_cache
    return cuda.compile_with_cache(
cupy/cuda/compiler.py:434: in compile_with_cache
    cache_in_memory, jitify)
cupy/cuda/compiler.py:510: in _compile_with_cache_cuda
    log_stream, cache_in_memory, jitify)
cupy/cuda/compiler.py:268: in compile_using_nvrtc
    name_expressions, log_stream, jitify)
cupy/cuda/compiler.py:244: in _compile
    source, options, cu_path)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

source = '/tmp/tmpnj_5wzp3/f72aa37c54d2e84fcc92a654e415e0a1_2.cubin.cu\n#include <cupy/complex.cuh>\n#include <cupy/carray.cuh>...ix_0 + ix_1])) * wval;\n        }\n        iws++;\n    }}\n    y = cast<Y>(sum);\n    ;\n      }\n      ;\n    }\n    '
options = ('--std=c++11', '-DCUPY_USE_JITIFY', '-I/home/lee8rx/my_git/pyir/cupy/cupy/core/include', '-I/usr/local/cuda/include', '-ftz=true', '-arch=compute_61')
cu_path = '/tmp/tmpnj_5wzp3/f72aa37c54d2e84fcc92a654e415e0a1_2.cubin.cu'

    def _jitify_prep(source, options, cu_path):
        # TODO(leofang): refactor this?
        global _jitify_header_source_map_populated
        if not _jitify_header_source_map_populated:
            from cupy.core import core
            _jitify_header_source_map = core._get_header_source_map()
            _jitify_header_source_map_populated = True
        else:
            # this is already cached at the C++ level, so don't pass in anything
            _jitify_header_source_map = None
    
        # jitify requires the 1st line to be the program name
        old_source = source
        source = cu_path + '\n' + source
    
        # Upon failure, in addition to throw an error Jitify also prints the log
        # to stdout. In principle we could intercept that by hijacking stdout's
        # file descriptor (tested locally), but the problem is pytest also does
        # the same thing internally, causing strange errors when running the tests.
        # As a result, we currently maintain Jitify's default behavior for easy
        # debugging, and wait for the upstream to address this issue
        # (NVIDIA/jitify#79).
    
        try:
            name, options, headers, include_names = jitify(
                source, options, _jitify_header_source_map)
        except Exception as e:  # C++ could throw all kinds of errors
            cex = CompileException(str(e), old_source, cu_path, options, 'jitify')
            dump = _get_bool_env_variable(
                'CUPY_DUMP_CUDA_SOURCE_ON_ERROR', False)
            if dump:
                cex.dump(sys.stderr)
>           raise JitifyException(str(cex))
E           AssertionError: Only cupy raises error
E           
E           Traceback (most recent call last):
E             File "/home/lee8rx/my_git/pyir/cupy/cupy/cuda/compiler.py", line 219, in _jitify_prep
E               source, options, _jitify_header_source_map)
E             File "cupy/cuda/jitify.pyx", line 59, in cupy.cuda.jitify.jitify
E               cpdef jitify(str code, tuple opt, dict cached_sources=None):
E             File "cupy/cuda/jitify.pyx", line 92, in cupy.cuda.jitify.jitify
E               load_program(cuda_source, headers, nullptr, &include_paths,
E           RuntimeError: Runtime compilation failed
E           
E           During handling of the above exception, another exception occurred:
E           
E           Traceback (most recent call last):
E             File "/home/lee8rx/my_git/pyir/cupy/cupy/testing/helper.py", line 47, in _call_func
E               result = impl(self, *args, **kw)
E             File "/home/lee8rx/my_git/pyir/cupy/tests/cupyx_tests/scipy_tests/ndimage_tests/test_filters.py", line 323, in test_filter
E               return self._filter(xp, scp)
E             File "/home/lee8rx/my_git/pyir/cupy/tests/cupyx_tests/scipy_tests/ndimage_tests/test_filters.py", line 89, in _filter
E               return filter(arr, *args, **kwargs)
E             File "/home/lee8rx/my_git/pyir/cupy/cupyx/scipy/ndimage/filters.py", line 320, in gaussian_filter1d
E               return correlate1d(input, weights, axis, output, mode, cval)
E             File "/home/lee8rx/my_git/pyir/cupy/cupyx/scipy/ndimage/filters.py", line 113, in correlate1d
E               return _correlate_or_convolve(input, weights, output, mode, cval, origins)
E             File "/home/lee8rx/my_git/pyir/cupy/cupyx/scipy/ndimage/filters.py", line 177, in _correlate_or_convolve
E               weights_dtype=weights_dtype)
E             File "/home/lee8rx/my_git/pyir/cupy/cupyx/scipy/ndimage/_filters_core.py", line 141, in _call_kernel
E               kernel(*args)
E             File "cupy/core/_kernel.pyx", line 821, in cupy.core._kernel.ElementwiseKernel.__call__
E               kern = self._get_elementwise_kernel(dev_id, arginfos, type_map)
E             File "cupy/core/_kernel.pyx", line 846, in cupy.core._kernel.ElementwiseKernel._get_elementwise_kernel
E               kern = _get_elementwise_kernel(
E             File "cupy/_util.pyx", line 59, in cupy._util.memoize.decorator.ret
E               result = f(*args, **kwargs)
E             File "cupy/core/_kernel.pyx", line 639, in cupy.core._kernel._get_elementwise_kernel
E               return _get_simple_elementwise_kernel(
E             File "cupy/core/_kernel.pyx", line 37, in cupy.core._kernel._get_simple_elementwise_kernel
E               cpdef function.Function _get_simple_elementwise_kernel(
E             File "cupy/core/_kernel.pyx", line 60, in cupy.core._kernel._get_simple_elementwise_kernel
E               module = compile_with_cache(module_code, options)
E             File "cupy/core/core.pyx", line 1986, in cupy.core.core.compile_with_cache
E               return cuda.compile_with_cache(
E             File "/home/lee8rx/my_git/pyir/cupy/cupy/cuda/compiler.py", line 434, in compile_with_cache
E               cache_in_memory, jitify)
E             File "/home/lee8rx/my_git/pyir/cupy/cupy/cuda/compiler.py", line 510, in _compile_with_cache_cuda
E               log_stream, cache_in_memory, jitify)
E             File "/home/lee8rx/my_git/pyir/cupy/cupy/cuda/compiler.py", line 268, in compile_using_nvrtc
E               name_expressions, log_stream, jitify)
E             File "/home/lee8rx/my_git/pyir/cupy/cupy/cuda/compiler.py", line 244, in _compile
E               source, options, cu_path)
E             File "/home/lee8rx/my_git/pyir/cupy/cupy/cuda/compiler.py", line 226, in _jitify_prep
E               raise JitifyException(str(cex))
E           cupy.cuda.compiler.JitifyException: Runtime compilation failed

cupy/cuda/compiler.py:226: AssertionError
------------------------------------------------------- Captured stdout call --------------------------------------------------------
---------------------------------------------------
--- JIT compile log for /tmp/tmpnj_5wzp3/f72aa37c54d2e84fcc92a654e415e0a1_2.cubin.cu ---
---------------------------------------------------
cupy/complex/complex.h(94): warning: __device__ annotation is ignored on a function("complex") that is explicitly defaulted on its first declaration

cupy/complex/complex.h(101): warning: __device__ annotation is ignored on a function("complex") that is explicitly defaulted on its first declaration

/tmp/tmpnj_5wzp3/f72aa37c54d2e84fcc92a654e415e0a1_2.cubin.cu(33): error: no operator "!=" matches these operands
            operand types are: W != int
          detected during instantiation of "__nv_bool nonzero(T) [with T=W]" 
(65): here

1 error detected in the compilation of "/tmp/tmpnj_5wzp3/f72aa37c54d2e84fcc92a654e415e0a1_2.cubin.cu".

---------------------------------------------------
------------------------------------------------------- Captured stderr call --------------------------------------------------------
```

Comment (leofang) [reply cupyx/scipy/ndimage/_filters_core.py:180]: Thanks for confirming and fixing it @grlee77! So this was in fact a mistake I made in #4264...I removed this specialization without realizing the complex support was not tested back then 😅

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by grlee77** [comment]
It looks like the [failure](https://jenkins.preferred.jp/job/chainer/job/cupy_pr/TEST=cupy-py36,label=mn1-p100/2775/console) here was a timeout? Probably, the changes to the source resulted in all ndimage filter kernels having to be recompiled.

**Comment by leofang** [comment]
Jenkins, test this please

File: cupyx/scipy/ndimage/_filters_core.py (hunk 1 above)
Comment (takagi) [reply cupyx/scipy/ndimage/_filters_core.py:180]: `complex` does not have a single-argument constructor so implicit conversion is not called.
The following with an explicit cast should work.

```cpp
template <class T>
__device__ __forceinline__ bool nonzero(T x) { return x != static_cast<T>(0); }
```

Comment (grlee77) [reply cupyx/scipy/ndimage/_filters_core.py]: Thanks @takagi, I confirmed locally that this works. I pushed a commit to the PR just now, but CI may have to run twice again due to potential time-outs.

Comment (leofang) [reply cupyx/scipy/ndimage/_filters_core.py]: Good catch, @takagi!

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by leofang** [comment]
Jenkins, test this please

**Comment by takagi** [comment]
LGTM, thanks!

**Comment by coderforlife** [comment]
@grlee77 I have a few comments/questions about this.

Now that it is merged I started testing with the signal functions that support complex values in scipy but have it blocked in cupyx waiting for complex support in `ndimage.convolve` since they all defer to that.

I started with `signal.bsplines.sepfir2d` and replaced the `if` statement for complex data types with:

```python
    if dtype.kind == 'c':
        dtype = cupy.complex64 if dtype == cupy.complex64 else cupy.complex128
```

At that point, complex data (with both real and imaginary parts non-zero) can be used for the `input` argument and you get the correct output as long as the `hrow` and `hcol` have zero-value imaginary parts. If there are imaginary values that are ones then it gets mostly the right value, but differences in the conjugation/signs of the parts. With other values in the imaginary part, the output is significantly off. However, even if they have zero-value imaginary parts the output values still have their imaginary parts modulated.

The `signal.convolve` and `signal.correlate` functions are a bit better off, but still not passing the tests. Additionally, they crash if provided one array that is complex and one that is float. This situation seems to be handled in your code before the call to `_get_correlate_kernel()` or calling the kernel directly (which is fine, just means something similar needs to be added to `signal._signaltools_core._direct_correlate()`).

Do you have any idea why this is occurring? What is different about your implementation of convolution/correlation with complex values and that used by `signal`? You said you just did the real and imaginary parts separately. Presumably not both input and the filter, but only one was separated (otherwise I wouldn't see the imaginary part change when zero in the filter). Do you have any idea how to incorporate after the fact?

**Comment by coderforlife** [comment]
Actually, I figured it out. Complex conjugation is needed in several cases depending on correlation vs convolution and if the inputs are swapped or not.
